nvtabular/io/parquet.py

```diff
@@ -677,17 +750,70 @@ def _close_writers(self):
             md_dict[fn] = writer.close(metadata_file_path=fn)
         return md_dict
 
-    def _bytesio_to_disk(self):
-        for bio, path in zip(self.data_bios, self.data_paths):
-            gdf = cudf.io.read_parquet(bio, index=False)
-            bio.close()
-            if self.shuffle == Shuffle.PER_WORKER:
-                gdf = _shuffle_gdf(gdf)
-            gdf.to_parquet(path, compression=None, index=False)
-        return
 
+class CPUParquetWriter(BaseParquetWriter):
+    def __init__(self, out_dir, **kwargs):
+        super().__init__(out_dir, **kwargs)
+        self.md_collectors = {}
+        self.pwriter_kwargs = {"compression": None}
+
+    @property
+    def _pwriter(self):
+        return pwriter_pyarrow
+
+    def _read_parquet(self, source):
+        return pd.read_parquet(source, engine="pyarrow")
+
+    def _to_parquet(self, df, sink):
+        md = []
+        df.to_parquet(sink, metadata_collector=md, compression=None, index=False)
+        fn = sink.split(self.fs.sep)[-1]
+        md[0].set_file_path(fn)
+        return md
+
+    def _append_writer(self, path, schema=None):
+
+        # Define "metadata collector" for pyarrow
+        _md_collector = []
+        _args = [schema]
+        _kwargs = {"metadata_collector": _md_collector}
+
+        # Use `BaseParquetWriter` logic
+        super()._append_writer(path, add_args=_args, add_kwargs=_kwargs)
+
+        # Keep track of "metadata collector" for pyarrow
+        self.md_collectors[path] = _md_collector
+
+    def _write_table(self, idx, data, has_list_column=False):
+        if has_list_column:
+            # Need to check if pandas supports chunked parquet writing
+            # with list columns.
+            filename = self._get_filename(len(self.data_paths))
+            data.to_parquet(filename)
+            self.data_paths.append(filename)
+        else:
+            table = pa.Table.from_pandas(data)
+            writer = self._get_or_create_writer(idx, schema=table.schema)
+            writer.write_table(table)
```

Comment:
Ah - We need to pass the desired number of rows in a row-group with the `row_group_size` argument here. Otherwise, we will get one large row-group per write.

Reply:
Okay - For now, we can just use `memory_usage` on the first partition processed by the `CPUParquetWriter` to approximate a "good" `row_group_size` argument (and cache the result for subsequent writes).